### DISCUSSION TITLE
Update AC First Timeslot

### DISCUSSION THREAD

**Discussion by allanwesleychang** [celsworth/lxp-bridge#170] [Ideas]
I think that it would be useful to include the ability to update the AC First start time and end times similarly to how the AC Charge timeslot can be adjusted in Home Assistant.

This would give one the ability to automate the switching between battery and grid.

**Comment by celsworth** [comment]
Hi!

I thought this was already done in https://github.com/celsworth/lxp-bridge/pull/143 - this lets you change the AC Charge / Charge Priority / Forced Discharge timeslots. These are the only three that I can see on Luxpowers portal.

Is one of those the one you mean by AC First, or is there another one we missed?

**Reply by allanwesleychang** [reply]
Hello,

I believe that it is one that was missed.

On my LuxpowerTek UI I'm referring to this field here:

![image](https://github.com/celsworth/lxp-bridge/assets/50325218/a5456313-3ce2-4fab-8653-c68e9f3d89fd)

Looking at [https://github.com/celsworth/lxp-bridge/blob/master/doc/LXP_REGISTERS.txt](url) it's register

> 152 AC_FIRST_START_HOUR / AC_FIRST_START_MINUTE
> 153 AC_FIRST_END_HOUR / AC_FIRST_END_MINUTE

Looking at [https://github.com/celsworth/lxp-bridge/blob/master/src/coordinator/commands/time_register_ops.rs](url)

```
fn register(&self) -> Result<i16> {
    use Action::*;
    match self {
        AcCharge(1) => Ok(68),
        AcCharge(2) => Ok(70),
        AcCharge(3) => Ok(72),
        ChargePriority(1) => Ok(76),
        ChargePriority(2) => Ok(78),
        ChargePriority(3) => Ok(80),
        ForcedDischarge(1) => Ok(84),
        ForcedDischarge(2) => Ok(86),
        ForcedDischarge(3) => Ok(88),
        _ => bail!("unsupported command"),
    }
}
```

I'm guessing that this code registers the different actions to the registers and 152 doesn't appear in the list.

Is my understanding correct?

**Reply by celsworth** [reply]
Ooh interesting, I don't see this AC First on my Luxpower portal - my inverter is getting pretty old now, and old firmware I guess, maybe thats why.

Yeah you seem to be correct, I think we can add this quite easily, its just a bit of copypaste with the new register numbers as far as I can see. I will take a look :)

**Reply by celsworth** [reply]
Made a start on this in #171. Couple of open questions:

What does this actually do, for the purposes of documentation?

Is there an "AC First Enable", like the "AC Charge Enable" button? If so we might need to find which register this is in so we can toggle that (otherwise the time ranges we set won't have any effect, much like for AC Charge already).

**Reply by allanwesleychang** [reply]
Thank you. Much appreciated.

There isn't an AC First Enable button. Basically, it sets the time frame where the inverter will use grid power over battery power. For example, if the time frame is set to 00h00 to 00h00, then the battery will be used to power the household instead of the grid. However, if it is set to 00h00 to 10h00 and the current time falls within this period then the grid will be used until the time elapses after which the battery will be used.

**Reply by celsworth** [reply]
I've merged the above PR so if you're in a position to try the `dev` Docker tag (or build from source), could you let me know how it goes?

I've not tested it but I'm reasonably sure it'll work as its mostly just a copy of AC Charge with some new register numbers.

**Reply by allanwesleychang** [reply]
I made use of the `dev` Docker images.

Initially, when viewing the controls in Home Assistant I saw:

![image](https://github.com/celsworth/lxp-bridge/assets/50325218/f8227a4c-9049-4b83-afe7-b9c7284e0e98)

not sure if `Unknown` is an issue but I did have values on the inverter set.

I set the first timeslot

![image](https://github.com/celsworth/lxp-bridge/assets/50325218/196632eb-ef7a-4057-9553-2b3dc255bdfd)

In the log I saw

> [2023-06-17T11:00:38.661 INFO lxp_bridge::mqtt] publishing: lxp/BA30803655/ac_first/1 = {"start":"00:00","end":"00:00"}
> [2023-06-17T11:00:38.662 INFO lxp_bridge::mqtt] publishing: lxp/result/BA30803655/set/ac_first/1 = OK

I can then confirm that the value was set currectly

![image](https://github.com/celsworth/lxp-bridge/assets/50325218/df9253dc-cae0-4a69-979e-0124fe8b9288)

I then reset the value

![image](https://github.com/celsworth/lxp-bridge/assets/50325218/a3231f96-580e-4006-91b4-390ec65859e7)

in the logs I saw

![image](https://github.com/celsworth/lxp-bridge/assets/50325218/bb7fa057-2434-4e9b-9889-c864a0e73f6e)

and then the value was set in the inverter stettings

![image](https://github.com/celsworth/lxp-bridge/assets/50325218/88945edc-e471-470c-b4b9-8189639f5e75)

So I can confirm that one is able to set the values.

Once again, thank you for making this update.

**Reply by celsworth** [reply]
Brilliant, thanks for testing. That all looks expected. The "unknown" values are because we don't currently broadcast the initial values to MQTT, so HA doesn't know what they are until we change them. There's an initiative by @lupine to improve this, but not sure on the timescale. It's a pretty minor issue though, mostly just cosmetic, doesn't affect functionality.

I'll get some docs added to the wiki and this will be in next release.

**Reply by lupine** [reply]
Hey,

Good to see someone other than me using this :sweat_smile:

We already publish all the individual registers at lxp-bridge startup; however, time registers are spread across several and home assistant doesn't know how to stitch them together. All we need to fix the behaviour here is another entry in this function: https://github.com/celsworth/lxp-bridge/blob/master/src/coordinator/mod.rs#L454

https://github.com/celsworth/lxp-bridge/issues/151 is the improved approach @celsworth is talking about, I think, but it's not really on my radar at the moment if I'm honest.

**Reply by lupine** [reply]
i've popped https://github.com/celsworth/lxp-bridge/pull/172 together in github's web edit UI; untested, but hopefully it compiles :sweat_smile:

**Reply by allanwesleychang** [reply]
Looks like that did the trick.

![image](https://github.com/celsworth/lxp-bridge/assets/50325218/a9276b12-ff05-4b27-be42-5e42e316a612)

**Comment by JohnMacrae** [comment]
You’re certainly not the only one! I would welcome somewhere where we can
discuss getting the best out of our systems. The HA community is a bit too
generic.


On Sat, 17 Jun 2023 at 14:26, Nick Thomas ***@***.***> wrote:

> Hey,
>
> Good to see someone other than me using this 😅
>
> We already publish all the individual registers at lxp-bridge startup;
> however, time registers are spread across several and home assistant
> doesn't know how to stitch them together. All we need to fix the behaviour
> here is another entry in this function:
> https://github.com/celsworth/lxp-bridge/blob/master/src/coordinator/mod.rs#L454
>
> #151 <https://github.com/celsworth/lxp-bridge/issues/151> is the improved
> approach @celsworth <https://github.com/celsworth> is talking about, I
> think, but it's not really on my radar at the moment if I'm honest.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/celsworth/lxp-bridge/discussions/170#discussioncomment-6204088>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADHTIDKTSR6KDY42BXAETMTXLWPABANCNFSM6AAAAAAZCQRK6Y>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***
> com>
>
-- 
Kind Regards
John
